### PR TITLE
Add unit tests for GDD.

### PR DESCRIPTION
**Pull request by fangzheng**
The unit tests use the standard cmockery framework. They complement
existing isolation2 tests for GDD, and allow us to test GDD with arbitrary wait graphs.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by ashwinstar** [comment]
Great! I didn't look at the code in details. But :+1: to the contribution

File: src/backend/utils/gdd/test/gdd_test.c [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+ *     and they wait for each other with solid edges on segments i-1
+ *     and i. E.g., txn 1 waits for txn 101 on segment 0, and txn 101
+ *     waits for txn 1 on segment 1, both with dotted edges.
+ *   - In total, there are 200 dotted edges and no deadlock.
+ */
+static void
+test_reduce_large_graph_no_deadlock1(void **state)
+{
+	const int num_transactions = 200;
+	TestWaitRelation wait_relations[num_transactions];
```
Comment (kainwen) [inline src/backend/utils/gdd/test/gdd_test.c:324]: When trying your code, I get the compile error:

```
gdd_test.c: In function ‘test_reduce_large_graph_pair_deadlocks’:
gdd_test.c:230:2: error: ISO C90 forbids variable length array ‘wait_relations’ [-Werror=vla]
  230 |  TestWaitRelation wait_relations[num_transactions];
      |  ^~~~~~~~~~~~~~~~
gdd_test.c: In function ‘test_reduce_large_graph_no_deadlock1’:
gdd_test.c:324:2: error: ISO C90 forbids variable length array ‘wait_relations’ [-Werror=vla]
  324 |  TestWaitRelation wait_relations[num_transactions];
      |  ^~~~~~~~~~~~~~~~
gdd_test.c: In function ‘test_reduce_large_graph_single_deadlock’:
gdd_test.c:417:2: error: ISO C90 forbids variable length array ‘wait_relations’ [-Werror=vla]
  417 |  TestWaitRelation wait_relations[num_transactions+1];
      |  ^~~~~~~~~~~~~~~~
gdd_test.c: In function ‘test_reduce_large_graph_no_deadlock2’:
gdd_test.c:495:2: error: ISO C90 forbids variable length array ‘wait_relations’ [-Werror=vla]
  495 |  TestWaitRelation wait_relations[num_transactions+1];
      |  ^~~~~~~~~~~~~~~~
```

Would you please help avoid this style?

Comment (fangzheng) [reply src/backend/utils/gdd/test/gdd_test.c]: Good catch. Changed to palloc() the arrays.

**Comment by kainwen** [comment]
LGTM.

I am asking if this contains in ci, once get the reply, will get this in.

Thanks a lot.

**Comment by asimrp** [comment]
I'm not seeing the newly added tests in the unittest output in this Travis job: https://api.travis-ci.com/v3/job/480094250/log.txt  It is surprising because the newly added Makefile looks correct.

File: src/backend/utils/gdd/test/gdd_test.c [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+ *-------------------------------------------------------------------------
+ */
+
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include "cmockery.h"
+
+#include "postgres.h"
+#include "../gdddetector.h"
```
Comment (asimrp) [inline src/backend/utils/gdd/test/gdd_test.c:32]: Unit-tests typically target a single .c file to test.  This particular test targets `gdddetector.c` file.  The .c file under test should be included in the unit-test.  My suggestion is therefore to replace this line with:
```C
#include gdddetector.c
```
Please refer to `src/backend/fts/test/*.c` for examples.  In that vein, the test file should also be renamed to `gdddetector_test.c`.

**Review by asimrp — COMMENTED** [review]
Thank you @fangzheng for this contribution.  The patch is very well documented and the tests it adds are quite meaningful.  A few small things need to be taken care of, as noted ~above~ below.

File: src/backend/utils/gdd/test/gdd_test.c (hunk 2 above)
Comment (fangzheng) [reply src/backend/utils/gdd/test/gdd_test.c]: Hi Asim,
Thanks a lot for the excellent suggestion. I've changed the test as you advised.

**Comment by asimrp** [comment]
Merged to master as fda1e439c84c1a54fbc3530652f49ab121766406.  Thanks again for the contribution, @fangzheng

**Pull request closed without merging**